File: binarysearch/Counting-Maximal-Value-Roots-in-Binary-Tree/Counting-Maximal-Value-Roots-in-Binary-Tree.cpp
// https://helloacm.com/counting-maximal-value-roots-in-binary-tree-using-depth-first-search-algorithm/
// https://binarysearch.com/problems/Counting-Maximal-Value-Roots-in-Binary-Tree
// EASY, DFS

/**
 * class Tree {
 *     public:
 *         int val;
 *         Tree *left;
 *         Tree *right;
 * };
 */
int solve(Tree* root) {
    int ret = 0;
    function<int(Tree*)> dfs = [&](Tree* root) {
        if (!root) return 0;
        int left = dfs(root->left);
        int right = dfs(root->right);
        if (root->val >= max(left, right)) {
            ret ++;
        }
        return max({root->val, left, right});
    };
   dfs(root);    
   return ret;
}

File: binarysearch/Counting-Maximal-Value-Roots-in-Binary-Tree/Counting-Maximal-Value-Roots-in-Binary-Tree.py
# https://helloacm.com/teaching-kids-programming-counting-maximal-value-roots-in-binary-tree-recursive-post-order-traversal-dfs-algorithm/
# https://binarysearch.com/problems/Counting-Maximal-Value-Roots-in-Binary-Tree
# EASY, DFS

# class Tree:
#     def __init__(self, val, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root):
        ans = 0

        def dfs(root):
            if not root:
                return -math.inf
            nonlocal ans
            left = dfs(root.left)
            right = dfs(root.right)
            if root.val >= left and root.val >= right:
                ans += 1
            return max(left, right, root.val)

        dfs(root)
        return ans
 
